client/relay-chain-rpc-interface/src/tokio_platform.rs

```diff
@@ -0,0 +1,327 @@
+// Copyright 2023 Parity Technologies (UK) Ltd.
+// This file is part of Cumulus.
+
+// Cumulus is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Cumulus is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Cumulus.  If not, see <http://www.gnu.org/licenses/>.
+
+use core::time::Duration;
+use futures::{channel::mpsc, prelude::*, task::Poll};
+use smoldot::libp2p::{multiaddr::ProtocolRef, websocket, Multiaddr};
+use smoldot_light::platform::{
+	ConnectError, PlatformConnection, PlatformSubstreamDirection, ReadBuffer,
+};
+use std::{
+	collections::VecDeque,
+	io::IoSlice,
+	net::{IpAddr, SocketAddr},
+	pin::Pin,
+	sync::Arc,
+};
+use tokio::net::TcpStream;
+
+use tokio_util::compat::TokioAsyncReadCompatExt;
+pub struct TokioPlatform;
+
+pub struct TokioStream {
+	shared: Arc<StreamShared>,
+	read_data_rx: Arc<parking_lot::Mutex<stream::Peekable<mpsc::Receiver<Vec<u8>>>>>,
+	read_buffer: Option<Vec<u8>>,
+}
+
+struct StreamShared {
+	guarded: parking_lot::Mutex<StreamSharedGuarded>,
+	write_queue_pushed: event_listener::Event,
+}
+
+struct StreamSharedGuarded {
+	write_queue: VecDeque<u8>,
+}
+
+/// Platform implementation for tokio
+/// This implementation is a conversion of the implementation for async-std:
+/// https://github.com/smol-dot/smoldot/blob/54d88891b1da202b4bf612a150df7b4dbfa03a55/light-base/src/platform/async_std.rs#L40
+impl smoldot_light::platform::Platform for TokioPlatform {
```

Comment:
> I would return the question and ask why rewrite this code for tokio when smoldot already provides something that works and is actively maintained.

In general I think it makes sense for us to standardize on a single async executor, and like it or not, everything runs on tokio these days.

Is there any particular reason why you're only supporting async-std and not tokio in smoldot? Considering that tokio is 1) faster than async-std, 2) more actively maintained (last updated 9 days ago vs last updated 12 months ago), 3) effectively "won" as the defacto Rust async executor, and 4) even has an LTS branch?

Reply:
Yes I have a strong reason: using a tokio socket panics if you don't use a tokio executor. Using tokio sockets from within a library is fundamentally a hack, because it ties a hidden implementation detail of the library with an implementation detail of the code that uses it.

Given that a light client connects to 3 peers, performance **doesn't matter**. At all. I want to strongly emphasize that. If you had a list of potential performance improvements to make, ordered from most impactful to least impactful, using tokio rather than async-std is probably item number 5000.

As for maintenance, having no maintenance is a good thing. It works. In which world code that is modified every week is considered better than code that is modified every year? Should we stop using zlib because it's unmaintained? I really don't understand this argument.

Reply:
As a heads up I have also renamed `AsyncStdTcpWebSocket` to `DefaultPlatform`: https://github.com/smol-dot/smoldot/pull/675

Reply:
> Yes I have a strong reason: using a tokio socket panics if you don't use a tokio executor. Using tokio sockets from within a library is fundamentally a hack, because it ties a hidden implementation detail of the library with an implementation detail of the code that uses it.

Do you mean that ideally it should return an error instead of panicking, or do you mean that it should implicitly and silently spawn an executor behind your back?

Nevertheless, in smolldot you already accept the platform trait which implies which executor is used, so how is this a *hidden* implementation detail?

> Given that a light client connects to 3 peers, performance **doesn't matter**.

Yes, I know. I was mostly speaking *in general*.

> As for maintenance, having no maintenance is a good thing. It works. In which world code that is modified every week is considered better than code that is modified every year?

Well, sure, fundamentally I agree with you here, *if* the code in question can be considered finished/done/complete/bugfree.

> Should we stop using zlib because it's unmaintained?

Yes, because the API's horrible and it's still getting CVEs.

Reply:
> Do you mean that ideally it should return an error instead of panicking, or do you mean that it should implicitly and silently spawn an executor behind your back?

Ideally it would just work, which could in principle technically be done with the help of the `std`.

A solution that would make tokio less bad could be to explicitly pass as parameter some kind of `RuntimeContext` object when creating a socket. It would give a compile-time guarantee that creating the socket will reasonably succeed, which is my pet peeve with tokio.

> Nevertheless, in smoldot you already accept the platform trait which implies which executor is used, so how is this a hidden implementation detail?

What is hidden is the fact that tokio uses a thread-local storage internally. Calling `TcpSocket::new` reads from that internal thread-local storage. This is a hidden behavior.

Unless I name the executor `PlatformThatRequiresYouToExecuteTheClientFromWithinATokioRuntime`, I will consider this a footgun. I don't consider it obvious for people that the word "tokio" implies "tokio must be used in your main application as well".